### PR TITLE
NODE-1374 Implement final revision of transactions (core)

### PR DESCRIPTION
**Pull request by mbroadst**
This is the core component to the changes required to pass the current revision of the transactions specification YAML tests.

### CODE REVIEW

File: lib/sessions.js [hunk 1]
```diff
@@ -22,6 +24,94 @@ function assertAlive(session, callback) {
   return true;
 }
 
+// Client session states
+const TxnState = {
```
Comment (daprahamian) [inline lib/sessions.js]: Can we refactor this into something that uses an IIFE to make it less verbose?

I was thinking something like:

```js
let TxnState;
let txnStateTransition;
(() => {
  const NO_TRANSACTION = 'NO_TRANSACTION';
  const STARTING_TRANSACTION = 'STARTING_TRANSACTION';
  const TRANSACTION_IN_PROGRESS: 'TRANSACTION_IN_PROGRESS';
  const TRANSACTION_COMMITTED: 'TRANSACTION_COMMITTED';
  const TRANSACTION_COMMITTED_EMPTY: 'TRANSACTION_COMMITTED_EMPTY';
  const TRANSACTION_ABORTED: 'TRANSACTION_ABORTED';

  const stateMachine = {...};

  TxnState = { NO_TRANSACTION, STARTING_TRANSACTION, TRANSACTIOKN_IN_PROGRESS ... };
  txnStateTransition = function(txnState, nextState) {
  };
})();
```

Comment (daprahamian) [reply lib/sessions.js]: Also, this is adding 114 lines to this file. Can we put it in a separate file?

File: lib/sessions.js [hunk 2]
```diff
@@ -219,73 +308,111 @@ class ClientSession extends EventEmitter {
...
-      callback(new MongoError('No transaction started'));
+  // handle any initial problematic cases
+  let txnState = session.transaction.state;
+
+  if (txnState === TxnState.NO_TRANSACTION) {
+    callback(new MongoError('No transaction started'));
+    return;
+  }
+
+  if (commandName === 'commitTransaction') {
```
Comment (daprahamian) [inline lib/sessions.js]: Is this a `const` somewhere?

File: lib/wireprotocol/3_2_support.js [hunk 3]
```diff
@@ -346,15 +368,17 @@ WireProtocol.prototype.command = function(bson, ns, cmd, cursorState, topology,
   } else if (cmd) {
     query = setupCommand(bson, ns, cmd, cursorState, topology, options);
   } else {
-    throw new MongoError(f('command %s does not return a cursor', JSON.stringify(cmd)));
+    return new MongoError(f('command %s does not return a cursor', JSON.stringify(cmd)));
```
Comment (daprahamian) [inline lib/wireprotocol/3_2_support.js]: Can we use template strings here?

File: lib/sessions.js [hunk 4]
```diff
@@ -219,73 +308,111 @@ class ClientSession extends EventEmitter {
...
-        return clientSession.topology.command(
-          'admin.$cmd',
-          command,
-          { session: clientSession },
-          (_err, _reply) => commandHandler(transactionError(_err), _reply)
-        );
-      }
-
-      commandHandler(transactionError(err), reply);
+  session.topology.command('admin.$cmd', command, { session }, (err, reply) => {
```
Comment (daprahamian) [inline lib/sessions.js]: This makes me feel weird for some reason, but I can't figure out a good way to refactor this.

I'd love for there to be some way to not write the `session.topology.command` out more than once.

Comment (mbroadst) [reply lib/sessions.js]: I'm going to leave this be, but certainly submit a PR if you come up with something you think is better!

**Pull request merged**